ts-client/src/dlmm/index.ts

```diff
@@ -5226,126 +5228,47 @@ export class DLMM {
-          const [lowerBinId, upperBinId] =
-            getBinArrayLowerUpperBinId(lowerBinArrayIndex);
-
-          const binArrayBins: Bin[] = [];
-          for (let i = lowerBinId.toNumber(); i <= upperBinId.toNumber(); i++) {
-            binArrayBins.push({
-              amountX: new BN(0),
-              amountY: new BN(0),
-              liquiditySupply: new BN(0),
-              rewardPerTokenStored: [new BN(0), new BN(0)],
-              amountXIn: new BN(0),
-              amountYIn: new BN(0),
-              feeAmountXPerTokenStored: new BN(0),
-              feeAmountYPerTokenStored: new BN(0),
-              price: new BN(0),
-            });
-          }
-
-          return {
-            bins: binArrayBins,
-            index: lowerBinArrayIndex,
-            version: 1,
-          };
-        }));
+    const hasCachedLowerBinArray = lowerBinArray != null;
+    const hasCachedUpperBinArray = upperBinArray != null;
+    const isSingleBinArray = lowerBinArrayIndex.eq(upperBinArrayIndex);
 
-      const [lowerBinIdForBinArray] = getBinArrayLowerUpperBinId(
-        binArray.index
-      );
+    const lowerBinArrayIndexOffset = hasCachedLowerBinArray ? 1 : 0;
+    const upperBinArrayIndexOffset = hasCachedUpperBinArray ? -1 : 0;
 
-      binArray.bins.forEach((bin, idx) => {
-        const binId = lowerBinIdForBinArray.toNumber() + idx;
-
-        if (binId >= lowerBinId && binId <= upperBinId) {
-          const pricePerLamport = getPriceOfBinByBinId(
-            binId,
-            this.lbPair.binStep
-          ).toString();
-          bins.push({
-            binId,
-            xAmount: bin.amountX,
-            yAmount: bin.amountY,
-            supply: bin.liquiditySupply,
-            price: pricePerLamport,
-            version: binArray.version,
-            pricePerToken: new Decimal(pricePerLamport)
-              .mul(new Decimal(10 ** (baseTokenDecimal - quoteTokenDecimal)))
-              .toString(),
-          });
-        }
-      });
-    } else {
-      const [lowerBinArrayPubKey] = deriveBinArray(
-        lbPairPubKey,
-        lowerBinArrayIndex,
-        this.program.programId
-      );
-      const [upperBinArrayPubKey] = deriveBinArray(
-        lbPairPubKey,
-        upperBinArrayIndex,
-        this.program.programId
-      );
-
-      const binArrays = await (async () => {
-        if (!lowerBinArrays || !upperBinArrays) {
-          return (
-            await this.program.account.binArray.fetchMultiple([
-              lowerBinArrayPubKey,
-              upperBinArrayPubKey,
-            ])
-          ).filter((binArray) => binArray !== null);
-        }
-
-        return [lowerBinArrays, upperBinArrays];
-      })();
+    const binArrayPubkeys = range(
+      lowerBinArrayIndex.toNumber() + lowerBinArrayIndexOffset,
+      upperBinArrayIndex.toNumber() + upperBinArrayIndexOffset,
+      i => deriveBinArray(lbPairPubKey, new BN(i), this.program.programId)[0]
+    );
```

Comment:
Need to have extra logic here, i think this will face issue when the differences bewteen`lowerBinArrayIndex` & `upperBinArrayIndex` is within 3. 
E.g:
- lowerBinArrayIndex = 1, upperBinArrayIndex = 1
- lowerBInArrayIndex = 1, upperBinArrayIndex = 2

the idea is to prevent `fetchMultiple` when binArray index range is within 2 because I believe most of the time, user would only interact with 2 bin array.

Reply:
If the diff is 0 (one bin array requested) it will return an array of one key to fetch.
If the diff is 1 it will return an array of two keys to fetch.
In a case when some cached bin arrays are provided the offsets are calculated to excluded excessive fetches.

E.g.:
```ts
range(1, 1, i => i) // [1]
range(1, 2, i => i) // [1, 2]
```

Reply:
ideally, we wan to avoid `fetchMultiple` in these condition. But returning `[1]` or `[1,2]` will result in `fetchMultiple`

Reply:
> ideally, we wan to avoid `fetchMultiple` in these condition. But returning `[1]` or `[1,2]` will result in `fetchMultiple`

Not sure what's the goal.
`fetchMultiple([1, 2])` will have the same perf impact as `fetch(1); fetch(2)`

Reply:
> ideally, we wan to avoid `fetchMultiple` in these condition. But returning `[1]` or `[1,2]` will result in `fetchMultiple`

Can handle it this way:
```ts
    const fetchedBinArrays = await (async () => {
      if (binArrayPubkeys.length === 0) {
        return [];
      } else if (binArrayPubkeys.length === 1) {
        const binArray0 = await this.program.account.binArray.fetchNullable(binArrayPubkeys[0]);
        return [binArray0];
      } else if (binArrayPubkeys.length === 2) {
        const binArray0 = await this.program.account.binArray.fetchNullable(binArrayPubkeys[0]);
        const binArray1 = await this.program.account.binArray.fetchNullable(binArrayPubkeys[2]);
        return [binArray0, binArray1];
      } else {
        return await this.program.account.binArray.fetchMultiple(binArrayPubkeys);
      }
    })();
```

Still not sure it's worth it.
Or you mean something else?

Reply:
Sorry for the late reply, what i mean is if there's only 2 bin array involved, there shouldn't be any fetching so that client side can save up some bandwidth in rpc call